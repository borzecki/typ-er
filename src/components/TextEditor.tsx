import React, { useState, useEffect } from "react";
import classnames from "classnames";

import useKeyPress from "../hooks/useKeyPress";

const Letter = ({ char, state }: { char: string; state: any }) => (
  <span className={classnames(state, { space: char === " " })}>{char}</span>
);

const TextEditor = ({
  text,
  resetTimer,
  increaseErrors,
  onComplete,
}: {
  text: string;
  resetTimer?: () => void;
  increaseErrors?: () => void;
  onComplete: () => void;
}) => {
  const key = useKeyPress();
  const [currentIndex, setIndex] = useState(0);
  const [failed, setFailed] = useState<Set<number>>(new Set([]));

  useEffect(() => {
    if (currentIndex === text.length) {
      onComplete();
      setFailed(new Set());
      setIndex(0);
    }
    if (key === "") return;

    // reset timer when first character is typed
    if (currentIndex === 0 && !failed.has(currentIndex) && resetTimer)
      resetTimer();

    if (key === text[currentIndex]) {
      setIndex(currentIndex + 1);
    } else {
      failed.add(currentIndex);
      setFailed(failed);
      increaseErrors && increaseErrors();
    }
    // eslint-disable-next-line
  }, [key]);

  const getState = (index: number) => {
    return {
      current: index === currentIndex,
      failed: failed.has(index),
      success: index < currentIndex && !failed.has(index),
      unused: index > currentIndex,
    };
  };

  return (
    <div className="editor">
      {[...text].map((char, index) => (
        <Letter key={index} char={char} state={getState(index)} />
      ))}
    </div>
  );
};

export default TextEditor;
