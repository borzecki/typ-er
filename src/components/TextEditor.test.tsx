import React from "react";
import { mount, ReactWrapper } from "enzyme";

import TextEditor from "./TextEditor";
import { act } from "react-dom/test-utils";

// simulates key down and key up actions with rerenders inbetween
const simulateKeyStroke = (
  key: string,
  map: { [x: string]: any; keydown?: any; keyup?: any },
  wrapper: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>
) => {
  act(() => {
    map.keydown({ key });
  });
  wrapper.update();

  act(() => {
    map.keyup();
  });
  wrapper.update();
};

describe("TextEditor tests", () => {
  test("completes the edit flow", () => {
    const map: { [x: string]: any } = {};
    window.addEventListener = jest.fn((event: string, cb) => {
      map[event] = cb;
    });
    const onComplete = jest.fn();
    const wrapper = mount(<TextEditor text="test" onComplete={onComplete} />);
    simulateKeyStroke("t", map, wrapper);
    simulateKeyStroke("e", map, wrapper);
    simulateKeyStroke("s", map, wrapper);
    simulateKeyStroke("t", map, wrapper);
    expect(onComplete).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  test("handles errors", () => {
    const map: { [x: string]: any } = {};
    window.addEventListener = jest.fn((event: string, cb) => {
      map[event] = cb;
    });
    const onComplete = jest.fn();
    const wrapper = mount(<TextEditor text="test" onComplete={onComplete} />);
    simulateKeyStroke("x", map, wrapper);
    simulateKeyStroke("t", map, wrapper);
    simulateKeyStroke("n", map, wrapper);
    simulateKeyStroke("Shift", map, wrapper);
    expect(onComplete).toHaveBeenCalledTimes(0);
    expect(wrapper).toMatchSnapshot();
  });
});
