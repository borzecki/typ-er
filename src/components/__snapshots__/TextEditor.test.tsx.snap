// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextEditor tests completes the edit flow 1`] = `
<TextEditor
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  text="test"
>
  <div
    className="editor"
  >
    <Letter
      char="t"
      key="0"
      state={
        Object {
          "current": true,
          "failed": false,
          "success": false,
          "unused": false,
        }
      }
    >
      <span
        className="current"
      >
        t
      </span>
    </Letter>
    <Letter
      char="e"
      key="1"
      state={
        Object {
          "current": false,
          "failed": false,
          "success": false,
          "unused": true,
        }
      }
    >
      <span
        className="unused"
      >
        e
      </span>
    </Letter>
    <Letter
      char="s"
      key="2"
      state={
        Object {
          "current": false,
          "failed": false,
          "success": false,
          "unused": true,
        }
      }
    >
      <span
        className="unused"
      >
        s
      </span>
    </Letter>
    <Letter
      char="t"
      key="3"
      state={
        Object {
          "current": false,
          "failed": false,
          "success": false,
          "unused": true,
        }
      }
    >
      <span
        className="unused"
      >
        t
      </span>
    </Letter>
  </div>
</TextEditor>
`;

exports[`TextEditor tests handles errors 1`] = `
<TextEditor
  onComplete={[MockFunction]}
  text="test"
>
  <div
    className="editor"
  >
    <Letter
      char="t"
      key="0"
      state={
        Object {
          "current": false,
          "failed": true,
          "success": false,
          "unused": false,
        }
      }
    >
      <span
        className="failed"
      >
        t
      </span>
    </Letter>
    <Letter
      char="e"
      key="1"
      state={
        Object {
          "current": true,
          "failed": true,
          "success": false,
          "unused": false,
        }
      }
    >
      <span
        className="current failed"
      >
        e
      </span>
    </Letter>
    <Letter
      char="s"
      key="2"
      state={
        Object {
          "current": false,
          "failed": false,
          "success": false,
          "unused": true,
        }
      }
    >
      <span
        className="unused"
      >
        s
      </span>
    </Letter>
    <Letter
      char="t"
      key="3"
      state={
        Object {
          "current": false,
          "failed": false,
          "success": false,
          "unused": true,
        }
      }
    >
      <span
        className="unused"
      >
        t
      </span>
    </Letter>
  </div>
</TextEditor>
`;
